src/Type/Definition/ArgumentsTrait.php

```diff
@@ -23,36 +42,38 @@ public function hasArguments(): bool
 
     /**
      * @return Argument[]
+     * @throws InvariantException
      */
     public function getArguments(): array
     {
         return $this->arguments;
     }
 
     /**
-     * @param array|Argument[] $arguments
-     * @return $this
+     * @param string $typeName
+     * @param array  $rawArguments
+     * @return Argument[]
      * @throws InvariantException
      */
-    protected function buildArguments(array $arguments)
+    protected function buildArguments(string $typeName, array $rawArguments): array
     {
         invariant(
-            isAssocArray($arguments),
-            'Args must be an associative array with argument names as keys.'
+            isAssocArray($rawArguments),
+            \sprintf(
+                '%s.%s args must be an object with argument names as keys.',
+                $typeName,
+                $this->getName()
+            )
         );
 
-        foreach ($arguments as $argumentName => $argument) {
-            $this->arguments[] = $argument instanceof Argument
-                ? $argument
-                : new Argument(
-                    $argumentName,
-                    $argument['type'] ?? null,
-                    $argument['defaultValue'] ?? null,
-                    $argument['description'] ?? null,
-                    $argument['astNode'] ?? null
-                );
+        $arguments = [];
+
+        foreach ($rawArguments as $argumentName => $argumentConfig) {
+            $argumentConfig['name'] = $argumentName;
+
+            $arguments[] = newArgument($argumentConfig);
```

Comment:
After cleaning up the logic in `ExtensionContext` I could simplify this to always assume that arguments are given in the "raw" format (associative array).